pyomo/contrib/observer/model_observer.py

```diff
@@ -0,0 +1,1294 @@
+                self._vars[v] = (lb, ub, fixed, domain_interval, value)
+                if reason & Reason.bounds:
+                    self._update_var_bounds(v)
+
+    def update_variables(self, variables: Optional[Collection[VarData]] = None):
+        self._update_variables(variables)
+        self._updates.run()
+
+    def _update_parameters(self, params: Optional[Collection[ParamData]] = None):
+        if params is None:
+            params = self._params
+        for p in params:
+            _val = self._params[p]
+            val = p.value
+            reason = Reason.no_change
+            if _val != val:
+                reason |= Reason.value
+            if reason:
+                self._updates.params_to_update[p] |= reason
+                self._params[p] = val
+
+    def update_parameters(self, params: Optional[Collection[ParamData]]):
+        self._update_parameters(params)
+        self._updates.run()
+
+    def _update_con(self, con: ConstraintData):
+        self._active_constraints[con] = con.expr
+        (named_exprs, variables, parameters, external_functions) = (
+            collect_components_from_expr(con.expr)
+        )
+        if named_exprs:
+            self._named_expressions[con] = [(e, e.expr) for e in named_exprs]
+        else:
+            self._named_expressions.pop(con, None)
+        if external_functions:
+            self._external_functions[con] = external_functions
+        else:
+            self._external_functions.pop(con, None)
+
+        _variables = self._vars_referenced_by_con[con]
+        _parameters = self._params_referenced_by_con[con]
+        new_vars = variables - _variables
+        old_vars = _variables - variables
+        new_params = parameters - _parameters
+        old_params = _parameters - parameters
+
+        self._vars_referenced_by_con[con] = variables
+        self._params_referenced_by_con[con] = parameters
+
+        if new_vars:
+            self._check_for_new_vars(new_vars)
+        if new_params:
+            self._check_for_new_params(new_params)
+
+        for v in new_vars:
+            self._referenced_variables[v][0].add(con)
+        for v in old_vars:
+            self._referenced_variables[v][0].remove(con)
+        for p in new_params:
+            self._referenced_params[p][0].add(con)
+        for p in old_params:
+            self._referenced_params[p][0].remove(con)
+
+        if old_vars:
+            self._check_to_remove_vars(old_vars)
+        if old_params:
+            self._check_to_remove_params(old_params)
+
+    def _update_constraints(self, cons: Optional[Collection[ConstraintData]] = None):
+        if cons is None:
+            cons = self._active_constraints
+        for c in cons:
+            reason = Reason.no_change
+            if c.expr is not self._active_constraints[c]:
+                reason |= Reason.expr
+            if reason:
+                self._updates.cons_to_update[c] |= reason
```

Comment:
Can't this be simplified to:
```suggestion
            if c.expr is not self._active_constraints[c]:
                self._updates.cons_to_update[c] |= Reason.expr
```